src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -335,4 +359,139 @@ public static Polygon getPixelGeometry (int x, int y, int zoom) {
                 new Coordinate(minLon, minLat)
         });
     }
+
+    /** Create grids from a CSV file */
+    public static Map<String,Grid> fromCsv(File csvFile, String latField, String lonField, int zoom) throws IOException {
+        CsvReader reader = new CsvReader(new BufferedInputStream(new FileInputStream(csvFile)), Charset.forName("UTF-8"));
+        reader.readHeaders();
+
+        String[] headers = reader.getHeaders();
+        if (!Stream.of(headers).filter(h -> h.equals(latField)).findAny().isPresent()) {
+            LOG.info("Lat field not found!");
+            halt(400, "Lat field not found");
+        }
+
+        if (!Stream.of(headers).filter(h -> h.equals(lonField)).findAny().isPresent()) {
+            LOG.info("Lon field not found!");
+            halt(400, "Lon field not found");
+        }
+
+        Envelope envelope = new Envelope();
+
+        // Keep track of which fields contain numeric values
+        Set<String> numericColumns = Stream.of(headers).collect(Collectors.toCollection(HashSet::new));
+        numericColumns.remove(latField);
+        numericColumns.remove(lonField);
+
+        int i = 0;
+        while (reader.readRecord()) {
+            if (++i % 10000 == 0) LOG.info("{} records", human(i));
+
+            envelope.expandToInclude(parseDouble(reader.get(lonField)), parseDouble(reader.get(latField)));
+
+            for (Iterator<String> it = numericColumns.iterator(); it.hasNext();) {
+                String field = it.next();
+                String value = reader.get(field);
+                if (value == null || "".equals(value)) continue; // allow missing data
+                try {
+                    // TODO also exclude columns containing negatives?
+                    parseDouble(value);
+                } catch (NumberFormatException e) {
+                    it.remove();
```

Comment:
Ah, never mind. The columns are removed the first time a non-numeric value is encountered. That means only a few exceptions will occur.